src/jupyterlabcontroller/storage/kubernetes/watcher.py

```diff
@@ -0,0 +1,221 @@
+"""Watch a Kubernetes namespace or cluster for events."""
+
+from __future__ import annotations
+
+from collections.abc import AsyncIterator, Awaitable, Callable
+from dataclasses import dataclass
+from datetime import timedelta
+from typing import Any, Generic, Self, TypeVar
+
+from kubernetes_asyncio.client import ApiException
+from kubernetes_asyncio.watch import Watch
+from structlog.stdlib import BoundLogger
+
+from ...exceptions import KubernetesError
+from ...models.domain.kubernetes import WatchEventType
+
+#: Type of Kubernetes object being watched (`dict` for custom objects).
+T = TypeVar("T")
+
+__all__ = [
+    "KubernetesWatcher",
+    "T",
+    "WatchEvent",
+]
+
+
+@dataclass
+class WatchEvent(Generic[T]):
+    """Parsed event from a Kubernetes watch.
+
+    This model is intended only for use within the Kubernetes storage layer.
+    It is interpreted by the callers of the generic watch class inside the
+    per-object Kubernetes storage classes.
+    """
+
+    action: WatchEventType
+    """Action the event represents."""
+
+    object: T
+    """Affected Kubernetes object."""
+
+    @classmethod
+    def from_event(cls, event: dict[str, Any], object_type: type[T]) -> Self:
+        """Create a `KubernetesWatchEvent` from a watch event.
+
+        Parameters
+        ----------
+        event
+            Event as returned by the Kubernetes watch API.
+        object_type
+            Expected type of the object.
+
+        Raises
+        ------
+        ValueError
+            Raised if the type of the object in the watch event was incorrect.
+        """
+        action = WatchEventType(event["type"])
+        if object_type.__name__ == "dict":
+            return cls(action=action, object=event["raw_object"])
+        obj = event["object"]
+        if not isinstance(obj, object_type):
+            real_type = type(obj).__name__
+            expected_type = object_type.__name__
+            msg = f"Watch object was of type {real_type}, not {expected_type}"
+            raise ValueError(msg)
+        return cls(action=action, object=obj)
+
+
+class KubernetesWatcher(Generic[T]):
+    """Watch Kubernetes for events.
+
+    This wrapper around the watch API of the Kubernetes client implements
```

Comment:
Eesh.  It feels like the low-effort upstreamable fix here would be to add an optional parameter to kubernetes_asyncio: if it's None, use the existing scary behavior of parsing the docstring, but if it's specified, trust the caller and use that.

No, I don't know if this would actually be low-effort.

Reply:
That is indeed how kubernetes_asyncio works. If you pass a type to the `Watch` constructor, then it uses that type; otherwise, it does the docstring parsing. That's one of the things that ths wrapper does: pass the explicit type into the constructor.